examples/tv-casting-app/darwin/MatterTvCastingBridge/MatterTvCastingBridge/MCEndpoint.mm

```diff
@@ -0,0 +1,127 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MCEndpoint_Internal.h"
+
+#import "MCCastingPlayer_Internal.h"
+#import "MCCluster_Internal.h"
+#import "MCDeviceTypeStruct.h"
+
+#import "MCClusterObjects.h"
+
+#import "clusters/Clusters.h"
+#import "core/Endpoint.h"
+
+#import <Foundation/Foundation.h>
+
+@interface MCEndpoint ()
+
+@property (nonatomic, readwrite) matter::casting::memory::Strong<matter::casting::core::Endpoint> cppEndpoint;
+
+@end
+
+@implementation MCEndpoint
+
+- (instancetype _Nonnull)initWithCppEndpoint:(matter::casting::memory::Strong<matter::casting::core::Endpoint>)cppEndpoint
+{
+    if (self = [super init]) {
+        _cppEndpoint = cppEndpoint;
+    }
+    return self;
+}
+
+- (NSNumber * _Nonnull)identifier
+{
+    return [NSNumber numberWithUnsignedShort:_cppEndpoint->GetId()];
+}
+
+- (NSNumber * _Nonnull)productId
+{
+    return [NSNumber numberWithUnsignedShort:_cppEndpoint->GetProductId()];
+}
+
+- (NSNumber * _Nonnull)vendorId
+{
+    return [NSNumber numberWithUnsignedShort:_cppEndpoint->GetVendorId()];
+}
+
+- (NSArray * _Nonnull)deviceTypeList
+{
+    NSMutableArray * deviceTypeList = [NSMutableArray new];
+    std::vector<chip::app::Clusters::Descriptor::Structs::DeviceTypeStruct::DecodableType> cppDeviceTypeList = _cppEndpoint->GetDeviceTypeList();
+    for (chip::app::Clusters::Descriptor::Structs::DeviceTypeStruct::DecodableType cppDeviceTypeStruct : cppDeviceTypeList) {
+        MCDeviceTypeStruct * deviceTypeStruct = [[MCDeviceTypeStruct alloc] initWithDeviceType:cppDeviceTypeStruct.deviceType revision:cppDeviceTypeStruct.revision];
+        [deviceTypeList addObject:deviceTypeStruct];
+    }
+    return deviceTypeList;
+}
+
+- (MCCastingPlayer * _Nonnull)castingPlayer
+{
+    return [[MCCastingPlayer alloc] initWithCppCastingPlayer:std::shared_ptr<matter::casting::core::CastingPlayer>(_cppEndpoint->GetCastingPlayer())];
+}
+
+- (MCCluster * _Nullable)clusterForType:(MCEndpointClusterType)type
+{
+    switch (type) {
+    case MCEndpointClusterTypeContentLauncher:
+        return [[MCContentLauncherCluster alloc] initWithCppCluster:_cppEndpoint->GetCluster<matter::casting::clusters::content_launcher::ContentLauncherCluster>()];
+
+    default:
+        ChipLogError(AppServer, "MCEndpointClusterType not found");
+        break;
+    }
+    return nil;
+}
+
+- (BOOL)hasCluster:(MCEndpointClusterType)type
+{
+    return [self clusterForType:type] != nil;
+}
+
+- (BOOL)isEqualToMCEndpoint:(MCEndpoint * _Nullable)other
+{
+    return [self.identifier isEqualToNumber:other.identifier];
```

Comment:
don't we need to compare productId and vendorId too? What if two MCEndpoints happen to share the same ID?

Reply:
EndpointId must be unique between two endpoints on the same casting player for casting to work E2E.